File: src/main/java/com/example/MonaServer/Entities/Group.java
package com.example.MonaServer.Entities;

import com.example.MonaServer.DTO.GroupDTO;
import com.example.MonaServer.Helper.ImageHelper;
import com.example.MonaServer.Helper.SecurityFilter;
import lombok.Getter;
import lombok.Setter;
import org.hibernate.annotations.OnDelete;
import org.hibernate.annotations.OnDeleteAction;

import javax.persistence.*;
import java.util.HashSet;
import java.util.Set;

@Entity(name = "groups")
@Getter
@Setter
public class Group {
    @Id
    @GeneratedValue(strategy = GenerationType.SEQUENCE, generator = "group_id_generator")
    @SequenceGenerator(name="group_id_generator", sequenceName = "group_id_seq", allocationSize = 1)
    @Column(name = "group_id", nullable = false)
    private Long groupId;

    @Column(name = "name", nullable = false, unique = true)
    private String name;

    @OneToOne
    @JoinColumn(name = "group_admin", nullable = false, referencedColumnName = "username")
    private User groupAdmin;

    @Column(name = "description")
    private String description;

    @Column(name = "profile_image", nullable = false)
    private byte[] profileImage;


    @Column(name = "pin_image", nullable = false)
    private byte[] pinImage;

    @Column(name = "invite_url", unique = true)
    private String inviteUrl;

    //TODO somthing like:
    // 0: public
    // 1 : private visible
    // 2 : private only invite
    @Column(name = "visibility", nullable = false)
    private int visibility;

    @ManyToMany(fetch = FetchType.LAZY,
            cascade = CascadeType.REMOVE,
            targetEntity = User.class)
    @JoinTable(
            name = "members",
            joinColumns = { @JoinColumn(name = "group_id") },
            inverseJoinColumns = { @JoinColumn(name = "username") }
    )
    @OnDelete(action= OnDeleteAction.CASCADE)
    private Set<User> members = new HashSet<>();

    @ManyToMany(fetch = FetchType.LAZY,
            cascade = CascadeType.REMOVE,
            targetEntity = Pin.class)
    @JoinTable(
            name = "groups_pins",
            joinColumns = { @JoinColumn(name = "group_id") },
            inverseJoinColumns = { @JoinColumn(name = "id"),  }
    )
    @OnDelete(action= OnDeleteAction.CASCADE)
    private Set<Pin> pins = new HashSet<>();

    public Group(String name, User groupAdmin, int visibility, String description, byte[] profileImage, byte[] pinImage) {
        this.name = name;
        this.groupAdmin = groupAdmin;
        this.visibility = visibility;
        this.description = description;
        this.profileImage = ImageHelper.getProfileImage(profileImage);
        this.pinImage = ImageHelper.getPinImage(profileImage);
        if (visibility != 0) {
            inviteUrl = SecurityFilter.generateAlphabeticRandomString(10);
        }
        this.members.add(groupAdmin);
        this.pinImage = pinImage;
    }

    public Group(GroupDTO groupDTO, User groupAdmin) {
        this.name = groupDTO.getName();
        this.groupAdmin = groupAdmin;
        this.visibility = groupDTO.getVisibility();
        this.description = groupDTO.getDescription();
        this.profileImage = ImageHelper.getProfileImage(groupDTO.getProfileImage());
        this.pinImage = ImageHelper.getPinImage(groupDTO.getProfileImage());
        if (visibility != 0) {
            inviteUrl = SecurityFilter.generateAlphabeticRandomString(10);
        }
        this.members.add(groupAdmin);
    }

    public Group() {}

    public void addGroupMember(User user) {
        members.add(user);
    }
    public void addPin(Pin pin) {
        pins.add(pin);
    }
    public void removeGroupMember(String username) {members.removeIf(e -> e.getUsername().equals(username));}
    public void updateGroupImage(byte[] image) {
        if (image != null) {
            this.profileImage = ImageHelper.getProfileImage(image);
            this.pinImage = ImageHelper.getPinImage(image);
        }
    }

    public void setInvite() {
        if (visibility != 0) {
            inviteUrl = SecurityFilter.generateAlphabeticRandomString(10);
        } else {
            inviteUrl = null;
        }
    }

}

File: src/main/kotlin/de/lrprojects/monaserver/entity/Group.kt
package de.lrprojects.monaserver.entity

import com.fasterxml.jackson.annotation.JsonIgnore
import de.lrprojects.monaserver.entity.keys.PreDeleteEntity
import de.lrprojects.monaserver.helper.SecurityHelper
import de.lrprojects.monaserver.properties.DbConstants.ADMIN_ID
import de.lrprojects.monaserver.properties.DbConstants.GROUP
import de.lrprojects.monaserver.properties.DbConstants.GROUPS
import de.lrprojects.monaserver.properties.DbConstants.ID
import de.lrprojects.monaserver.properties.DbConstants.MEMBER_GROUP
import de.lrprojects.monaserver.types.DeletedEntityType
import jakarta.persistence.CascadeType
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.FetchType
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.OneToMany
import jakarta.persistence.Table
import jakarta.persistence.Transient
import org.hibernate.annotations.CreationTimestamp
import org.hibernate.annotations.UpdateTimestamp
import java.time.OffsetDateTime
import java.util.Objects
import javax.sql.DataSource


@Entity
@Table(name = GROUPS)
data class Group (

    @Column(nullable = false, unique = true)
    var name: String? = null,

    @ManyToOne(fetch = FetchType.EAGER)
    @JoinColumn(name = ADMIN_ID, referencedColumnName = ID)
    var groupAdmin: User? = null,

    @Column
    var description: String? = null,

    @Column(unique = true)
    var inviteUrl: String? = null,

    @Column
    var link: String? = null,

    @Column(nullable = false)
    var visibility: Int = 0,

    @OneToMany(mappedBy = MEMBER_GROUP, fetch = FetchType.LAZY, cascade = [CascadeType.REMOVE], orphanRemoval = true)
    @JsonIgnore
    var members: MutableSet<Member> = mutableSetOf(),

    @OneToMany(mappedBy = ID, fetch = FetchType.LAZY, cascade = [CascadeType.REMOVE], orphanRemoval = true)
    @JsonIgnore
    var pins: MutableSet<Pin> = HashSet(),

    @OneToMany(mappedBy = GROUP, fetch = FetchType.LAZY, cascade = [CascadeType.REMOVE], orphanRemoval = true)
    @JsonIgnore
    var seasons: MutableSet<GroupSeason> = HashSet(),

    @CreationTimestamp
    @Column
    var creationDate: OffsetDateTime? = null,

    @UpdateTimestamp
    @Column
    var updateDate: OffsetDateTime? = null,

    @Column(nullable = false)
    var isDeleted: Boolean = false,

    @Transient
    private var dataSource: DataSource? = null

): PreDeleteEntity() {
    fun setInvite() {
        inviteUrl = if (visibility != 0) {
            SecurityHelper.generateAlphabeticRandomString(10)
        } else {
            null
        }
    }

    override fun getDeletedEntityType() = DeletedEntityType.GROUP
    override fun equals(other: Any?): Boolean {
        if (this === other) return true
        if (other !is Group) return false

        if (id != null && other.id != null) {
            return id == other.id
        }

        return name == other.name &&
                groupAdmin?.id == other.groupAdmin?.id
    }

    override fun hashCode(): Int {
        return id?.hashCode() ?: name?.hashCode() ?: 0
    }


}
